ts: Implement new on-disk format for time series

mrtracy opened the pull request:
The first commit is #25587 and can be ignored for this PR.


Implement a new *columnar* on-disk format for time series samples,
replacing the previous row-like format.

Previously, each slab of time series data contained a collection of
"samples", where each sample was a message containing a number of
fields; the timestamp (offset), and a number of value fields intended to
contain "rolled-up" data for long-term, low resolution storage.

The new format is columnar; the top-level slab contains multiple
parallel arrays, with each array containing the ordered values for the
individual samples.

This gives us the following advantages:

This has a columnar layout, made up of parallel arrays. This gives us
all of the advantages we were missing in the previous layout:

+ High-resolution data can leave the aggregate fields completely empty;
only the "last" aggregate fields. This will reduce the in-memory size of
each sample from the full Sample structure down to a int32 and a
float64. This also means we can add several more aggregates without
inflating the size of each sample.
+ The columnar format takes advantage of protobuffer repeated field
packing, which should save considerable space for the encoded on-disk
format.
+ When querying, we can iterate directly over the data fields we need,
which may improve data locality (and thus cache-miss performance) or
allow us to more aggressively release memory for data that is not
needed.

This commit does the following:

+ Define new columnar fields in roachpb/internal.proto.
+ Write new C++ merge logic for the columnar fields. This does not
replace the existing row logic; it lives alongside it.
+ Create an upgrade path in the merge logic for row-formatted slabs;
this occurs whenever columnar data is merged into an existing key with
row-formatted data. This ensures than any individual slab of data will
contain only row-formatted samples or column-formatted samples, but
never both.

Release note: none.

Commit:
ts: Implement new on-disk format for time series

Implement a new *columnar* on-disk format for time series samples,
replacing the previous row-like format.

Previously, each slab of time series data contained a collection of
"samples", where each sample was a message containing a number of
fields; the timestamp (offset), and a number of value fields intended to
contain "rolled-up" data for long-term, low resolution storage.

The new format is columnar; the top-level slab contains multiple
parallel arrays, with each array containing the ordered values for the
individual samples.

This gives us the following advantages:

This has a columnar layout, made up of parallel arrays. This gives us
all of the advantages we were missing in the previous layout:

+ High-resolution data can leave the aggregate fields completely empty;
only the "last" aggregate fields. This will reduce the in-memory size of
each sample from the full Sample structure down to a int32 and a
float64. This also means we can add several more aggregates without
inflating the size of each sample.
+ The columnar format takes advantage of protobuffer repeated field
packing, which should save considerable space for the encoded on-disk
format.
+ When querying, we can iterate directly over the data fields we need,
which may improve data locality (and thus cache-miss performance) or
allow us to more aggressively release memory for data that is not
needed.

This commit does the following:

+ Define new columnar fields in roachpb/internal.proto.
+ Write new C++ merge logic for the columnar fields. This does not
replace the existing row logic; it lives alongside it.
+ Create an upgrade path in the merge logic for row-formatted slabs;
this occurs whenever columnar data is merged into an existing key with
row-formatted data. This ensures than any individual slab of data will
contain only row-formatted samples or column-formatted samples, but
never both.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25789)
<!-- Reviewable:end -->

petermattis commented:
I only did a pass on the C++ code. We have support for C++ unit tests now. I'd like to see some added for `convertToColumnar` and `sortAndDeduplicateColumns`.

---

Review status: 0 of 19 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[c-deps/libroach/merge.cc, line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7hZtd9VmiyVxst62Q:-LD7hZtd9VmiyVxst62R:b-23ql2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L98)):*
> ```cpp
> }
> 
> void sortAndDeduplicateColumns(cockroach::roachpb::InternalTimeSeriesData& data, int first_unsorted) {
> ```

The C++ style I'm used to says that mutable parameters should be passed as a pointers, not as non-const references. This makes it more obvious at the caller that the parameter will be mutated. `s/& data/* data/g`. This comment applies above and possibly other places in this PR.

---

*[c-deps/libroach/merge.cc, line 105 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7hN4P8KarNJpb8s1D:-LD7hN4P8KarNJpb8s1E:b7gsbfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L105)):*
> ```cpp
>   auto order = std::vector<int>(data.offset_size() - first_unsorted);
>   std::iota(order.begin(), order.end(), first_unsorted);
>   std::stable_sort(order.begin(), order.end(), [&](const int &a, const int &b) {
> ```

A const reference to an integer is usually worse than an integer as it implies a level of indirection that isn't needed. So `s/const int&/const int/g`. This comment also applies a few places below.

---

*[c-deps/libroach/merge.cc, line 130 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7iaFhD3_-JO-qKm9F:-LD7iaFhD3_-JO-qKm9G:b-s2u7ht) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L130)):*
> ```cpp
>     if (src_idx == dest_idx) {
>       continue;
>     }
> ```

Did you get this from somewhere? It is subtle. 

---

*[c-deps/libroach/merge.cc, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7iwh19dQcLlROXGqA:-LD7iwh19dQcLlROXGqB:bm507dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L215)):*
> ```cpp
>       left_ts.offset().begin(),
>       std::lower_bound(left_ts.offset().begin(), left_ts.offset().end(), *min_offset)
>     );
> ```

The `first_unsorted_index` stuff is an optimization, right? It would be safe to always set `first_unsorted_index` to `0`. How important do you think this is?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-:-LD7jJOa9FDHHBka9NIP:bg3ojsb)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Added C++ tests and addressed comments.

---

Review status: 0 of 11 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/merge.cc, line 98 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7hZtd9VmiyVxst62Q:-LDDA6Ih1jAG7JlGklFZ:b-uw48vx) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L98)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The C++ style I'm used to says that mutable parameters should be passed as a pointers, not as non-const references. This makes it more obvious at the caller that the parameter will be mutated. `s/& data/* data/g`. This comment applies above and possibly other places in this PR.
</blockquote></details>

Done, there were two such new functions in this PR.

---

*[c-deps/libroach/merge.cc, line 105 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7hN4P8KarNJpb8s1D:-LDDAR3H0cumtThkN7D5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L105)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

A const reference to an integer is usually worse than an integer as it implies a level of indirection that isn't needed. So `s/const int&/const int/g`. This comment also applies a few places below.
</blockquote></details>

Done.

---

*[c-deps/libroach/merge.cc, line 130 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7iaFhD3_-JO-qKm9F:-LD7q--h66HkttLSayn_:bfjpa12) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L130)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you get this from somewhere? It is subtle. 
</blockquote></details>

No, just tried to figure out a way to swap these values in place.  I'll add some comments to the algorithm and some tests (although there already is a test for this in the Go code, which is just tests).

---

*[c-deps/libroach/merge.cc, line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LD7iwh19dQcLlROXGqA:-LD7sADf6eXiUYP_dM1D:bclxl18) ([raw file](https://github.com/cockroachdb/cockroach/blob/20664b9d534b9e399ade7dbce10ddf117671ca11/c-deps/libroach/merge.cc#L215)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The `first_unsorted_index` stuff is an optimization, right? It would be safe to always set `first_unsorted_index` to `0`. How important do you think this is?
</blockquote></details>

This is not a small optimization; in a merge, we can assume the entire left side is already sorted and deduplicated.  So, if we're at the end of an hour, and the hour already has 350 samples, and we're only merging in 10 new ones, this will only need to sort the last 10 samples.  Note that I am *also* accounting for the case where the unsorted samples on the right side overlap with the already-sorted left-side samples, which is why I don't just sort the right side and merge it in.  In that case, we might end up sorting (for example) the last 11 samples instead of the last 10.

If you have the Admin UI open, there are a lot of situations where there will be 1 correctly ordered sample being merged in.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-:-LDDg8j8CYn1PMgxQQqk:b5d624j)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 12 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/merge.h, line 30 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LDpzcmM3jGPAJ99yW5i:-LDpzcmM3jGPAJ99yW5j:b-bsuhpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e7f260a0c50ffdb4a5467bd0404c00f6ac92297/c-deps/libroach/merge.h#L30)):*
> ```objc
> DBStatus MergeResult(cockroach::storage::engine::enginepb::MVCCMetadata* meta, DBString* result);
> rocksdb::MergeOperator* NewMergeOperator();
> void sortAndDeduplicateColumns(roachpb::InternalTimeSeriesData* data, int first_unsorted);
> ```

Should there be comments here?

---

*[c-deps/libroach/merge.cc, line 211 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LDq-aHR4VlmsmpthL8L:-LDq-aHR4VlmsmpthL8M:b-1soj7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e7f260a0c50ffdb4a5467bd0404c00f6ac92297/c-deps/libroach/merge.cc#L211)):*
> ```cpp
>   }
> 
>   if (val_ts.offset_size() > 0) {
> ```

Add a comment what this means.

---

*[c-deps/libroach/merge.cc, line 339 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LDq0BRJBYnCb3SUjBBH:-LDq0BRJBYnCb3SUjBBI:bwzzm3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e7f260a0c50ffdb4a5467bd0404c00f6ac92297/c-deps/libroach/merge.cc#L339)):*
> ```cpp
> // data needs to be sorted.
> void sortAndDeduplicateColumns(cockroach::roachpb::InternalTimeSeriesData* data, int first_unsorted) {
>   // Create an auxilliary array of array indexes, and sort that array according
> ```

auxiliary

Have to interrupt my review here because of some user debugging. Sorry!

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25789)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 1 of 12 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[c-deps/libroach/merge.h, line 30 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LDpzcmM3jGPAJ99yW5i:-LDs-1dlBHTu57mUlTHE:b-lnu8v2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e7f260a0c50ffdb4a5467bd0404c00f6ac92297/c-deps/libroach/merge.h#L30)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Should there be comments here?
</blockquote></details>

I'm not sure, but there are substantial comments on the implementations, I think that is sufficient.

---

*[c-deps/libroach/merge.cc, line 211 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LDq-aHR4VlmsmpthL8L:-LDs-Za913mUbonU-s1m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e7f260a0c50ffdb4a5467bd0404c00f6ac92297/c-deps/libroach/merge.cc#L211)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Add a comment what this means.
</blockquote></details>

Done.

---

*[c-deps/libroach/merge.cc, line 339 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LDq0BRJBYnCb3SUjBBH:-LDs2e3uELrblf_73spK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e7f260a0c50ffdb4a5467bd0404c00f6ac92297/c-deps/libroach/merge.cc#L339)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

auxiliary

Have to interrupt my review here because of some user debugging. Sorry!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25789)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/merge.cc, line 372 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEES9DB6FhnQ-ZTiz_U:-LEES9DB6FhnQ-ZTiz_V:b9esitn) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/c-deps/libroach/merge.cc#L372)):*
> ```cpp
>     //
>     // Because we are swapping values, and because we iterate through
>     // destinations from front to back, it is possible that value that was
> ```

The explanation here is already really good, but I needed to write down an example. Might be worth inserting one like it.

```
data = [3 1 1 5]
order = [2 0 3]

i=0
data = [1 1 3 5]

i=1
order[1] = 0, dest_idx = 1
order[order[1]] = 2 corresponding to the index of `3` in `data`.
```

---

*[c-deps/libroach/merge_test.cc, line 89 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEXTqtAdXJT6NI4ld-:-LEEXTqtAdXJT6NI4ld0:b-qxhibw) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/c-deps/libroach/merge_test.cc#L89)):*
> ```cpp
>         {8, 1},
>         {9, 2},
>         {10, 0},
> ```

If you ran this test case with `first_unsorted=4`, it would duplicate the `10`, right? Is it an issue that `first_unsorted = n` also requires everything at `>= n` to have strictly larger offsets than `data[n-1]`?

---

*[c-deps/libroach/merge_test.cc, line 93 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEWaObDGcZmnpLilL6:-LEEWaObDGcZmnpLilL7:b-jc8zxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/c-deps/libroach/merge_test.cc#L93)):*
> ```cpp
>         {10, 3},
>       },
>       3,
> ```

This is a little unintuitive because the prefix isn't actually sorted, and so the result looks weird. Add a comment. (Sorting the prefix in the test weakens what you're testing, so I'm not suggesting you do that).

---

*[pkg/roachpb/internal.proto, line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYCilEzhrcmxAmWLS:-LEEYCilEzhrcmxAmWLT:b-jbhrec) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L51)):*
> ```protobuf
>   repeated InternalTimeSeriesSample samples = 3 [(gogoproto.nullable) = false, deprecated=true];
>   // Columnar array containing the ordered offsets of the samples in this
>   // data set. This column is always populated.
> ```

Does this comment (and the others) need to be clarified? It's always populated unless `samples` is also populated, right? Perhaps this qualifier could live on the message to avoid repeating it endlessly.

---

*[pkg/roachpb/internal.proto, line 64 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYfOM2rXRFZAuRUd_:-LEEYfONBB_X91pq7PqA:byk75oz) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L64)):*
> ```protobuf
>   // this sample period. This column is optional, and is only populated for
>   // samples which are "rollups" of multiple measurements; if missing, its value
>   // for all samples is 1.
> ```

Is this "if missing" in the sense of, "it's intentionally elided when it's equal to one because that's such a common case"? Why is it so common?

---

*[pkg/roachpb/internal.proto, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYnUs4mQR0dB4sLGd:-LEEYnUs4mQR0dB4sLGe:b4qqmcu) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L68)):*
> ```protobuf
>   // Columnar array containing the maximum value of any single measurement taken
>   // during this sample period. This column is optional, and is only populated
>   // for samples which are "rollups" of multiple measurements; ; if missing, its
> ```

Similar question, and spurious `;`.

---

*[pkg/roachpb/internal.proto, line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYpsI9gMLSiPmcjdX:-LEEYpsI9gMLSiPmcjdY:b-cjyn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L74)):*
> ```protobuf
>   // taken during this sample period. This column is optional, and is only
>   // populated for samples which are "rollups" of multiple measurements; if
>   // missing, its value for all samples is equal to "last".
> ```

Ditto here and below.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-:-LEE_5ol4UgHXl-tX1Qs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/merge.cc, line 372 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEES9DB6FhnQ-ZTiz_U:-LEFrSti0tEsVEr9xxoR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/c-deps/libroach/merge.cc#L372)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The explanation here is already really good, but I needed to write down an example. Might be worth inserting one like it.

```
data = [3 1 1 5]
order = [2 0 3]

i=0
data = [1 1 3 5]

i=1
order[1] = 0, dest_idx = 1
order[order[1]] = 2 corresponding to the index of `3` in `data`.
```
</blockquote></details>

Done.

---

*[c-deps/libroach/merge_test.cc, line 89 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEXTqtAdXJT6NI4ld-:-LEFvx171LCQn2H_aEp4:b86tj7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/c-deps/libroach/merge_test.cc#L89)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

If you ran this test case with `first_unsorted=4`, it would duplicate the `10`, right? Is it an issue that `first_unsorted = n` also requires everything at `>= n` to have strictly larger offsets than `data[n-1]`?
</blockquote></details>

You are correct; it actually already duplicates the "10", that's what the first data point is intended to illustrate. I have added a comment clarifying the construction of this test case and what it is intended to demonstrate.

---

*[c-deps/libroach/merge_test.cc, line 93 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEWaObDGcZmnpLilL6:-LEFseH4EL1_VKBYcNcq:bb3k2am) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/c-deps/libroach/merge_test.cc#L93)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This is a little unintuitive because the prefix isn't actually sorted, and so the result looks weird. Add a comment. (Sorting the prefix in the test weakens what you're testing, so I'm not suggesting you do that).
</blockquote></details>

Added a comment explaining what this is testing.

---

*[pkg/roachpb/internal.proto, line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYCilEzhrcmxAmWLS:-LEFupsaD3Yaq-VFdQJp:b-evxh75) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L51)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Does this comment (and the others) need to be clarified? It's always populated unless `samples` is also populated, right? Perhaps this qualifier could live on the message to avoid repeating it endlessly.
</blockquote></details>

Moved a lot of this information to the message comment.

---

*[pkg/roachpb/internal.proto, line 64 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYfOM2rXRFZAuRUd_:-LEFvsJlDRuUetsZa3vv:bldn04f) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L64)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Is this "if missing" in the sense of, "it's intentionally elided when it's equal to one because that's such a common case"? Why is it so common?
</blockquote></details>

I've expanded on this situation significantly in the message-level comment.

---

*[pkg/roachpb/internal.proto, line 68 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYnUs4mQR0dB4sLGd:-LEFuyT16eqlR5fYNPCG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L68)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Similar question, and spurious `;`.
</blockquote></details>

Done.

---

*[pkg/roachpb/internal.proto, line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/25789#-LEEYpsI9gMLSiPmcjdX:-LEFuy0P1VYRGfBOGSpI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09e402463f1ca2d8254e1414540e0778850f6c30/pkg/roachpb/internal.proto#L74)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Ditto here and below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25789)*
<!-- Sent from Reviewable.io -->
